src/Cache/Signature/FixerSignature.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Cache\Signature;
+
+use PhpCsFixer\Fixer\FixerInterface;
+
+final class FixerSignature
+{
+    private string $name;
+    private string $contentHash;
```

Comment:
I do believe that this is naive implementation and I am not feeling comfortable with it.

### problem 1
- imagine this hierarchy `AbstractFixer -> AbstractPhpDocFixer -> PhpDocReturnTypeFixer`
- have `contentHash` of `PhpDocReturnTypeFixer` , yet someone changed implementation of one of methods in `AbstractPhpDocFixer` - will `contentHash` change? I don't see that happening. And lack of `contentHash` change means that fixer will not run, even when it should.

### problem 2
imagine `FooFixer` using `BarService`, and we change `BarService` - again, `FooFixer` should run, but it won't.

### problem 3+
probably more to come

### solution?
we should likely get hash of everything `FooFixer` using, all the parent classes, traits, side classes, libraries, ...

Reply:
@keradus please have in mind that this PR addresses mainly 2 things: local development and custom fixers. For regular usage this PR does not change anything, since average users don't modify fixers (or their "components") in vendor dir, and basically the only time fixers change for those users is when Fixer's version is bumped, which invalidates cache anyway. The goal of this change is described above ([here](#issue-1729436920) and [here](#issuecomment-1566856031)).

Scout's rule:
- are custom fixers handled better with this changes? Yes, because upgrading custom fixers invalidates cache.
- is local development better with this changes? Yes, because cache can finally be used during development.
- is ad-hoc debugging better with this change? Yes, because when you modify code in `vendor` it will invalidate the cache.
- is it problem that in some scenarios cache won't be invalidated (because change will be made not in fixer class, but in some helper or parent class)? IMHO no, because it will be status quo, since now the cache wouldn't be invalidated too.

Not to mention cache can be disabled by the local config, if anyone want to keep current behavior (always full Fixer run). We can even add it to `.php-cs-fixer.dist.php` and disable cache by default.

Reply:
> For regular usage this PR does not change anything, since average users don't modify fixers (or their "components") in vendor dir,

For local development of PHP CS Fixer, contributor _is_ modifying fixers (or their components) in `src/` dir - and Fixer is not installed in vendor dir.
For development of any project that uses PHP CS Fixer as dependency, cache should work already without this PR.

What am I missing?

Reply:
> For local development of PHP CS Fixer, contributor is modifying fixers (or their components) in src/ dir - and Fixer is not installed in vendor dir.

Yes, but for local development cache was completely disabled. With this PR cache is allowed (can be disabled through config) and will be invalidated when fixers' code has changed. This improves running self-fixing on the codebase where you work on non-fixer classes - currently you need to run whole pass, with this PR cache will be used and only modified files will be fixed (or whole codebase, if fixers where modified).

Also, as I said before, it will improve support for custom fixers in terms of cache.

Reply:
I ignore custom fixer aspect till my concern about original repository is solved, to limit cognitive load of reviewing this PR.

What's the code that can be changed for original repo?
format: what changed -> should all rules be executed?
- one of rule changed -> yes
- tokenizer/* changed -> yes
- phpdoc parser changed -> yes

any of main repo part changed -> should run rules.
I prefer cache to be disabled when we cannot ensure that it's not needed to run the rules, than enable the case when we are only guessing if any change should be applied or not.
To my understanding, this proposal enables cache even if Tokenizer/* code is changed, and making Runner to not execute any rules when "only" Tokenizer/* changed since previous run - it's wrong behaviour.

imho instead of bringing falsy cache for self-repo execution (limited usage, approach proven to provide falsy cache values) we could look into other way of speeding up the work of Fixer (eg parallel run)

Reply:
mh...
https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/doc/usage.rst?plain=1#L145

Reply:
Is it actually a problem if for some runs with cache result would be a false success?
- it's developer's responsibility, if something more than fixers was changed, cache can be cleared (Composer script 👀) or skipped by CLI flag
- there's always CI that will catch it

Especially second point is something that we should focus on - optimise CI and make it super-reliable. This change is only for local optimisation, that can be more loose (just like coding and pushing changes without running tests, PHPStan etc.).

Reply:
for me it is a problem.

For me, the cache should be only helping, not giving me a false feeling that "all is great", and then CI fails.
The tool outcome should be the same, with or without cache. Cache existence should only improve performance, not affect the correctness of the results.

I like the part of "developers responsibility" - I would rather say to introduce the toggle for core contributors "enforce cache usage", rather than opting-in by default for a mechanism that we know will not work properly. or drop the idea fully till we figure out how to have it reliable.

Reply:
> For me, the cache should be only helping, not giving me a false feeling that "all is great", and then CI fails.
The tool outcome should be the same, with or without cache. Cache existence should only improve performance, not affect the correctness of the results.

Yet, we have `.php-cs-fixer.dist.php` which can be overriden with `.php-cs-fixer.php`, so if dist file is changed you can still run Fixer with outdated (local) config and have success, and then fail in the CI 🤷‍♂️ . Also, until recent Composer script introduction there wasn't any way to imitate full CI locally, even now it's hard to do it. My point is that you rather can't ensure that if something passes locally, it'll pass in CI too, because there are many little things that can impact it. That's why it's not an argument for me to shoot ourselves in the foot by disabling cache for local development when we're backed by CI and can catch local glitches later.

Especially that it can be easily disabled through the config, so if any of developers has problems with cache, then can opt-out from it and go back to previous behavior (or maybe we could disabled it by default in local development, to keep current behavior?).

Reply:
so, once more - as whole your msg is skipping this part.


> The tool outcome should be the same, with or without cache. Cache existence should only improve performance, not affect the correctness of the results.


to me, it is wrong to have cache that is making tool producing wrong outcome, giving user false feeling that "code is following CS properly".

With that shortcoming, I decided not to merge this PR.

I really like the idea behind it, and appreciate the effort put into the PR, yet the results is not showing stability I want to have to let it get merged.

Reply:
@keradus this is really, really weird stance that is protecting some ephemeral value at the cost of developer's time. For example in PHPStan with Symfony extension you can get false results both when:
- you don't run `cache:clear` command and you run analysis with wrong DI definition
- you clear DI cache after PHPStan's run, because result cache already contains errors related to missing services
Yet, [it is not a priority issue](https://github.com/phpstan/phpstan-symfony/issues/255) for Ondrej and I, as a developer, also can live with it - I just know that sometimes I need to remove result cache manually.

This PR has real value, especially in terms of custom fixers, but you prefer not to merge it because sometimes it can produce false result that would be caught anyway in the CI - I just don't get it 😩.